helpers/remoteoauth/token.go

```diff
@@ -0,0 +1,186 @@
+package remoteoauth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	cloudquery_api "github.com/cloudquery/cloudquery-api-go"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+)
+
+func NewTokenSource(opts ...TokenSourceOption) (oauth2.TokenSource, error) {
+	t := &tokenSource{}
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	if _, cloudEnabled := os.LookupEnv("CQ_CLOUD"); !cloudEnabled {
+		return oauth2.StaticTokenSource(&t.currentToken), nil
+	}
+
+	cloudToken, err := newCloudTokenSource()
+	if err != nil {
+		return nil, err
+	}
+	if t.noWrap {
+		return cloudToken, nil
+	}
+
+	return oauth2.ReuseTokenSource(nil, cloudToken), nil
+}
+
+type tokenSource struct {
+	currentToken oauth2.Token
+	noWrap       bool
+}
+
+type cloudTokenSource struct {
+	apiClient *cloudquery_api.ClientWithResponses
+	mu        sync.Mutex // protects against multiple refresh calls
+
+	apiURL           string
+	apiToken         string
+	teamName         string
+	syncName         string
+	testConnUUID     uuid.UUID
+	syncRunUUID      uuid.UUID
+	connectorUUID    uuid.UUID
+	isTestConnection bool
+}
+
+var _ oauth2.TokenSource = (*cloudTokenSource)(nil)
+
+func newCloudTokenSource() (oauth2.TokenSource, error) {
+	t := &cloudTokenSource{}
+
+	err := t.initCloudOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	t.apiClient, err = cloudquery_api.NewClientWithResponses(t.apiURL,
+		cloudquery_api.WithRequestEditorFn(func(_ context.Context, req *http.Request) error {
+			req.Header.Set("Authorization", "Bearer "+t.apiToken)
+			return nil
+		}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create api client: %w", err)
+	}
+
+	return t, nil
+}
+
+// Token returns the cached token if not expired, or a new token from the remote source.
+func (t *cloudTokenSource) Token() (*oauth2.Token, error) {
+	return t.TokenWithContext(context.TODO())
+}
+
+// TokenWithContext returns a new token from the remote source using the given context.
+func (t *cloudTokenSource) TokenWithContext(ctx context.Context) (*oauth2.Token, error) {
```

Comment:
This is not part of the TokenSource interface I think, and you shouldn't need the locking here since `reuseTokenSource` already has it.

If you need to pass the `context` I think you can do it as an option for `NewTokenSource`

Reply:
done in https://github.com/cloudquery/plugin-sdk/pull/1866/commits/2600aeebdc288826d9c850ca97f38a3fcc2cba4f